packages/ckeditor5-html-support/src/emptyblocks.ts

```diff
@@ -0,0 +1,186 @@
+/**
+ * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
+ */
+
+/**
+ * @module html-support/emptyblocks
+ */
+
+import { Plugin, type Editor } from 'ckeditor5/src/core.js';
+import type {
+	UpcastElementEvent,
+	Element,
+	DowncastDispatcher,
+	UpcastDispatcher,
+	DowncastAttributeEvent,
+	ElementCreatorFunction,
+	Node
+} from 'ckeditor5/src/engine.js';
+
+const EMPTY_BLOCK_MODEL_ATTRIBUTE = 'htmlEmptyBlock';
+
+/**
+ * This plugin allows for preserving empty block elements in the editor content instead of
+ * automatically filling them with block fillers (`&nbsp;`).
+ *
+ * This is useful when you want to:
+ *
+ *	* Preserve empty block elements exactly as they were in the source HTML
+ *	* Allow for styling empty blocks with CSS (block fillers can interfere with height/margin)
+ *	* Maintain compatibility with external systems that expect empty blocks to remain empty
+ *
+ * For example, this allows for HTML like:
+ *
+ * ```html
+ * <p></p>
+ * <p class="spacer"></p>
+ * <td></td>
+ * ```
+ * to remain empty instead of being converted to:
+ *
+ * ```html
+ * <p>&nbsp;</p>
+ * <p class="spacer">&nbsp;</p>
+ * <td>&nbsp;</td>
+ * ```
+ */
+export default class EmptyBlocks extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'EmptyBlocks' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static override get isOfficialPlugin(): true {
+		return true;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public afterInit(): void {
+		const { model, conversion } = this.editor;
+		const schema = model.schema;
+
+		schema.extend( '$block', {
+			allowAttributes: [ EMPTY_BLOCK_MODEL_ATTRIBUTE ]
+		} );
+
+		schema.extend( '$container', {
+			allowAttributes: [ EMPTY_BLOCK_MODEL_ATTRIBUTE ]
+		} );
+
+		// Downcasts.
+		conversion.for( 'dataDowncast' ).add( createEmptyBlocksDowncastDispatcher() );
+		conversion.for( 'editingDowncast' ).add( createEmptyBlocksDowncastDispatcher() );
+
+		// Upcasts.
+		conversion.for( 'upcast' ).add( createEmptyBlocksUpcastDispatcher( this.editor ) );
+
+		// Table related converters.
+		if ( schema.isRegistered( 'tableCell' ) ) {
+			schema.extend( 'tableCell', {
+				allowAttributes: [ EMPTY_BLOCK_MODEL_ATTRIBUTE ]
+			} );
+
+			conversion.for( 'dataDowncast' ).elementToElement( {
+				model: 'paragraph',
+				view: convertEmptyBlockParagraphInTableCell(),
+				converterPriority: 'highest'
+			} );
```

Comment:
This is overriding the default converter, can we handle this case similarly to other cases and after a feature handles it then we just set filler offset based on the final output? This way we would not need to duplicate part of paragraph downcast in a table cell.

Reply:
Done https://github.com/ckeditor/ckeditor5/pull/17756/commits/99f91c7f42db292e85acf85b9ffb68ab520927e6